File: sdk/go/lantern/client_test.go
package lantern

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
	"time"

	"github.com/zbloss/lantern/internal/domain"
)

// TestClient_GetPrompt_Label verifies GetPrompt fetches a prompt by label
// and returns the template.
func TestClient_GetPrompt_Label(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodGet {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}
		label := r.URL.Query().Get("label")
		if label == "" {
			label = "production"
		}
		pv := domain.PromptVersion{
			Name:      "greeting",
			Version:   1,
			Template:  "Hello, {{.Name}}!",
			ModelConfig: domain.PromptModelConfig{
				Model:       "gpt-4",
				Temperature: 0.7,
				MaxTokens:   100,
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(pv)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")

	template, err := client.GetPrompt("greeting", "production")
	if err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}
	if template != "Hello, {{.Name}}!" {
		t.Errorf("template: got %q, want %q", template, "Hello, {{.Name}}!")
	}
}

// TestClient_GetPrompt_DefaultLabel verifies label defaults to "production".
func TestClient_GetPrompt_DefaultLabel(t *testing.T) {
	var receivedLabel string
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		receivedLabel = r.URL.Query().Get("label")
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "test",
			Version:  1,
			Template: "test",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPrompt("test", "")
	if err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}
	if receivedLabel != "production" {
		t.Errorf("expected default label 'production', got %q", receivedLabel)
	}
}

// TestClient_GetPrompt_Version verifies GetPromptVersion fetches by version.
func TestClient_GetPrompt_Version(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		version := r.URL.Query().Get("version")
		if version == "" {
			http.Error(w, "version required", http.StatusBadRequest)
			return
		}
		pv := domain.PromptVersion{
			Name:     "greeting",
			Version:  2,
			Template: "Welcome, {{.Name}}!",
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(pv)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")

	template, err := client.GetPromptVersion("greeting", 2)
	if err != nil {
		t.Fatalf("GetPromptVersion: %v", err)
	}
	if template != "Welcome, {{.Name}}!" {
		t.Errorf("template: got %q, want %q", template, "Welcome, {{.Name}}!")
	}
}

// TestClient_GetPrompt_VersionNotFound verifies 404 returns an error.
func TestClient_GetPrompt_VersionNotFound(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "not found", http.StatusNotFound)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPromptVersion("nonexistent", 99)
	if err == nil {
		t.Error("expected error for nonexistent prompt, got nil")
	}
}

// TestClient_GetPrompt_Caching verifies label cache uses 30-second TTL.
func TestClient_GetPrompt_Caching(t *testing.T) {
	var requestCount atomic.Int64
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestCount.Add(1)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "cached",
			Version:  1,
			Template: "cached content",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")

	// First call — should hit the server.
	_, err := client.GetPrompt("cached", "production")
	if err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}

	// Second call — should use cache (no additional request within TTL).
	_, err = client.GetPrompt("cached", "production")
	if err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}

	// Within TTL, only 1 request.
	if requestCount.Load() != 1 {
		t.Errorf("expected 1 request (cached), got %d", requestCount.Load())
	}
}

// TestClient_GetPromptVersion_NoTTL verifies version cache has no TTL.
func TestClient_GetPromptVersion_NoTTL(t *testing.T) {
	var requestCount atomic.Int64
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requestCount.Add(1)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "immutable",
			Version:  1,
			Template: "immutable content",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")

	// First call — should hit the server.
	_, err := client.GetPromptVersion("immutable", 1)
	if err != nil {
		t.Fatalf("GetPromptVersion: %v", err)
	}

	// Second call — should use cache.
	_, err = client.GetPromptVersion("immutable", 1)
	if err != nil {
		t.Fatalf("GetPromptVersion: %v", err)
	}

	if requestCount.Load() != 1 {
		t.Errorf("expected 1 request (version cache, no TTL), got %d", requestCount.Load())
	}
}

// TestClient_GetPrompt_NotFound verifies 404 returns an error.
func TestClient_GetPrompt_NotFound(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "not found", http.StatusNotFound)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPrompt("nonexistent", "production")
	if err == nil {
		t.Error("expected error for nonexistent prompt, got nil")
	}
}

// TestClient_WriteScore verifies WriteScore sends a score to the server.
func TestClient_WriteScore(t *testing.T) {
	var receivedScore domain.ScoreRequest

	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}
		if err := json.NewDecoder(r.Body).Decode(&receivedScore); err != nil {
			http.Error(w, "invalid JSON", http.StatusBadRequest)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(map[string]string{"score_id": "score-123"})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")

	err := client.WriteScore("span-abc", "helpfulness", 0.8, "Great answer")
	if err != nil {
		t.Fatalf("WriteScore: %v", err)
	}

	if receivedScore.SpanID != "span-abc" {
		t.Errorf("span_id: got %q, want %q", receivedScore.SpanID, "span-abc")
	}
	if receivedScore.EvalName != "helpfulness" {
		t.Errorf("eval_name: got %q, want %q", receivedScore.EvalName, "helpfulness")
	}
	if receivedScore.Value != 0.8 {
		t.Errorf("value: got %f, want 0.8", receivedScore.Value)
	}
	if receivedScore.Reasoning != "Great answer" {
		t.Errorf("reasoning: got %q, want %q", receivedScore.Reasoning, "Great answer")
	}
}

// TestClient_WriteScore_RequiresSpanID verifies WriteScore fails for empty span ID.
func TestClient_WriteScore_RequiresSpanID(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "span_id is required", http.StatusBadRequest)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	err := client.WriteScore("", "eval", 1.0, "no span")
	if err == nil {
		t.Error("expected error for empty span ID")
	}
}

// TestClient_GetPrompt_ServerError verifies HTTP errors are returned.
func TestClient_GetPrompt_ServerError(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "internal error", http.StatusInternalServerError)
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPrompt("test", "production")
	if err == nil {
		t.Error("expected error for server error, got nil")
	}
}

// TestClient_BaseURL verifies the client sends requests to the correct baseURL.
func TestClient_BaseURL(t *testing.T) {
	var receivedPath string
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		receivedPath = r.URL.Path
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "test",
			Version:  1,
			Template: "test",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPrompt("test", "production")
	if err != nil {
		t.Fatalf("GetPrompt: %v", err)
	}
	if receivedPath != "/api/v1/prompts/test" {
		t.Errorf("path: got %q, want %q", receivedPath, "/api/v1/prompts/test")
	}
}

// TestClient_GetPromptVersion_BaseURL verifies version endpoint path.
func TestClient_GetPromptVersion_BaseURL(t *testing.T) {
	var receivedPath string
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		receivedPath = r.URL.Path
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "test",
			Version:  3,
			Template: "version 3",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPromptVersion("test", 3)
	if err != nil {
		t.Fatalf("GetPromptVersion: %v", err)
	}
	if receivedPath != "/api/v1/prompts/test" {
		t.Errorf("path: got %q, want %q", receivedPath, "/api/v1/prompts/test")
	}
}

// TestClient_SlowResponse verifies the client handles slightly slow responses
// without error (the 10-second timeout is generous enough for this 100ms delay).
func TestClient_SlowResponse(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(100 * time.Millisecond)
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(domain.PromptVersion{
			Name:     "test",
			Version:  1,
			Template: "test",
		})
	}))
	defer ts.Close()

	client := NewClient(ts.URL, "ltn_proj_test")
	_, err := client.GetPrompt("test", "production")
	if err != nil {
		t.Errorf("expected no error, got: %v", err)
	}
}

File: sdk/ts/tests/client.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { OmnevalClient } from "../src/client";
import { mockFetch, createResponse } from "./utils";

describe("OmnevalClient", () => {
  beforeEach(() => vi.clearAllMocks());
  afterEach(() => vi.restoreAllMocks());

  describe("getPrompt", () => {
    it("fetches a prompt by label", async () => {
      const fetchSpy = mockFetch(async (url) => {
        expect(url).toContain("/api/v1/prompts/greeting?label=production");
        return createResponse(200, {
          name: "greeting",
          version: 1,
          template: "Hello, {{.Name}}!",
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const template = await client.getPrompt("greeting", "production");

      expect(template).toBe("Hello, {{.Name}}!");
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("defaults label to production", async () => {
      const fetchSpy = mockFetch(async (url) => {
        expect(url).toContain("?label=production");
        return createResponse(200, {
          name: "test",
          version: 1,
          template: "test",
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.getPrompt("test");

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("caches prompt results for 30 seconds", async () => {
      const fetchSpy = mockFetch(async () => {
        return createResponse(200, {
          name: "cached",
          version: 1,
          template: "cached content",
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.getPrompt("cached", "production");
      await client.getPrompt("cached", "production");
      await client.getPrompt("cached", "production");

      // Only one fetch call due to caching
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("throws on 404", async () => {
      mockFetch(async () => {
        return createResponse(404);
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(
        client.getPrompt("nonexistent", "production")
      ).rejects.toThrow("prompt not found");
    });

    it("throws on server error", async () => {
      mockFetch(async () => {
        return createResponse(500, { error: "internal error" });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(
        client.getPrompt("test", "production")
      ).rejects.toThrow("get prompt: 500:");
    });

    it("fetches by version", async () => {
      const fetchSpy = mockFetch(async (url) => {
        expect(url).toContain("?version=2");
        return createResponse(200, {
          name: "v2",
          version: 2,
          template: "version 2 content",
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const template = await client.getPrompt("v2", { version: 2 });

      expect(template).toBe("version 2 content");
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("caches version results with no TTL", async () => {
      const fetchSpy = mockFetch(async () => {
        return createResponse(200, {
          name: "immutable",
          version: 1,
          template: "immutable content",
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.getPrompt("immutable", { version: 1 });
      await client.getPrompt("immutable", { version: 1 });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });
  });

  describe("writeScore", () => {
    it("writes a score", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.span_id).toBe("span-abc");
        expect(body.eval_name).toBe("helpfulness");
        expect(body.value).toBe(0.8);
        expect(body.reasoning).toBe("Great answer");
        return createResponse(201, { score_id: "score-123" });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.writeScore("span-abc", {
        name: "helpfulness",
        value: 0.8,
        reasoning: "Great answer",
      });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("throws on empty span ID", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(
        client.writeScore("", { name: "eval", value: 1.0 })
      ).rejects.toThrow("span_id is required");
    });

    it("throws on server error", async () => {
      mockFetch(async () => {
        return createResponse(500, { error: "internal error" });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(
        client.writeScore("span-1", { name: "eval", value: 1.0 })
      ).rejects.toThrow("write score: 500:");
    });

    it("includes API key header when configured", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        return createResponse(201);
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_proj_test",
      });
      await client.writeScore("span-1", { name: "eval", value: 1.0 });

      const call = fetchSpy.mock.calls[0];
      const headers = (call[1] as RequestInit)?.headers as Record<string, string>;
      expect(headers["X-API-Key"]).toBe("oev_proj_test");
    });

    it("includes project_id extracted from oev_proj_ api key in score payload", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.project_id).toBe("myprojectsuffix");
        return createResponse(201);
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_proj_myprojectsuffix",
      });
      await client.writeScore("span-1", { name: "eval", value: 1.0 });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("includes project_id extracted from oev_svc_ api key in score payload", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.project_id).toBe("myservicesuffix");
        return createResponse(201);
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_svc_myservicesuffix",
      });
      await client.writeScore("span-1", { name: "eval", value: 1.0 });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("uses full api key as project_id when key has no recognized prefix", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.project_id).toBe("raw-key");
        return createResponse(201);
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "raw-key",
      });
      await client.writeScore("span-1", { name: "eval", value: 1.0 });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("uses empty string as project_id when no api key configured", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.project_id).toBe("");
        return createResponse(201);
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.writeScore("span-1", { name: "eval", value: 1.0 });

      expect(fetchSpy).toHaveBeenCalledOnce();
    });
  });

  describe("createPrompt", () => {
    it("posts to /api/v1/prompts and returns prompt data", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.name).toBe("greeting");
        expect(body.template).toBe("Hello, {{.Name}}!");
        return createResponse(201, {
          name: "greeting",
          version: 1,
          template: "Hello, {{.Name}}!",
          model: "gpt-4",
          temperature: 0.7,
          max_tokens: 100,
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.createPrompt("greeting", "Hello, {{.Name}}!", {
        model_config: { model: "gpt-4", temperature: 0.7, max_tokens: 100 },
      });

      expect(result.name).toBe("greeting");
      expect(result.version).toBe(1);
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("sends X-API-Key header when configured", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const headers = (init?.headers as Record<string, string>) ?? {};
        expect(headers["X-API-Key"]).toBe("oev_proj_test");
        return createResponse(201, { name: "test", version: 1, template: "t" });
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_proj_test",
      });
      await client.createPrompt("test", "template");

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("throws on empty name", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.createPrompt("", "template")).rejects.toThrow("name is required");
    });

    it("throws on empty template", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.createPrompt("my-prompt", "")).rejects.toThrow("template is required");
    });

    it("throws on server error", async () => {
      mockFetch(async () => createResponse(409, { error: "version already exists" }));

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.createPrompt("greeting", "Hello")).rejects.toThrow("create prompt: 409:");
    });
  });

  describe("listPrompts", () => {
    it("returns list of prompt summaries", async () => {
      const fetchSpy = mockFetch(async (url) => {
        expect(url).toContain("/api/v1/prompts");
        return createResponse(200, [
          { name: "greeting", latest_version: 2, labels: { production: 2, staging: 1 } },
          { name: "eval", latest_version: 1, labels: {} },
        ]);
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.listPrompts();

      expect(result).toHaveLength(2);
      expect(result[0].name).toBe("greeting");
      expect(result[0].latest_version).toBe(2);
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("sends X-API-Key header when configured", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const headers = (init?.headers as Record<string, string>) ?? {};
        expect(headers["X-API-Key"]).toBe("oev_proj_test");
        return createResponse(200, []);
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_proj_test",
      });
      await client.listPrompts();

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("returns empty array when no prompts", async () => {
      mockFetch(async () => createResponse(200, []));

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.listPrompts();
      expect(result).toEqual([]);
    });
  });

  describe("setPromptLabel", () => {
    it("sends PUT to /api/v1/prompts/:name/labels/:label", async () => {
      const fetchSpy = mockFetch(async (url, init) => {
        expect(url).toContain("/api/v1/prompts/greeting/labels/production");
        expect(init?.method).toBe("PUT");
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.version).toBe(2);
        return createResponse(200, { name: "greeting", label: "production", version: 2 });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await client.setPromptLabel("greeting", "production", 2);

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("sends X-API-Key header when configured", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const headers = (init?.headers as Record<string, string>) ?? {};
        expect(headers["X-API-Key"]).toBe("oev_proj_test");
        return createResponse(200, { name: "test", label: "production", version: 1 });
      });

      const client = new OmnevalClient({
        baseUrl: "http://localhost:3000",
        apiKey: "oev_proj_test",
      });
      await client.setPromptLabel("test", "production", 1);

      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("throws on empty name", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.setPromptLabel("", "production", 1)).rejects.toThrow("name is required");
    });

    it("throws on server error", async () => {
      mockFetch(async () => createResponse(404, { error: "not found" }));

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.setPromptLabel("nonexistent", "production", 99)).rejects.toThrow(
        "set prompt label: 404:"
      );
    });
  });

  describe("listEvalRules", () => {
    it("returns list of eval rules", async () => {
      const fetchSpy = mockFetch(async (url) => {
        expect(url).toContain("/api/v1/eval-rules");
        return createResponse(200, {
          rules: [
            {
              rule_id: "rule-1",
              name: "helpfulness",
              judge_model: "gpt-4o-mini",
              prompt_name: "eval-prompt",
              prompt_version: 1,
              filter: {},
              sample_rate: 1.0,
              enabled: true,
            },
          ],
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.listEvalRules();

      expect(result).toHaveLength(1);
      expect(result[0].name).toBe("helpfulness");
      expect(result[0].rule_id).toBe("rule-1");
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("returns empty array when no rules", async () => {
      mockFetch(async () => createResponse(200, { rules: [] }));

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.listEvalRules();
      expect(result).toEqual([]);
    });
  });

  describe("createEvalRule", () => {
    it("posts to /api/v1/eval-rules and returns rule data", async () => {
      const fetchSpy = mockFetch(async (_url, init) => {
        const body = JSON.parse((init?.body as string) ?? "{}");
        expect(body.name).toBe("helpfulness");
        expect(body.prompt_name).toBe("eval-prompt");
        return createResponse(201, {
          rule_id: "rule-123",
          name: "helpfulness",
          judge_model: "gpt-4o-mini",
          prompt_name: "eval-prompt",
          prompt_version: 1,
          filter: {},
          sample_rate: 1.0,
          enabled: true,
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.createEvalRule("helpfulness", "eval-prompt");

      expect(result.rule_id).toBe("rule-123");
      expect(result.name).toBe("helpfulness");
      expect(fetchSpy).toHaveBeenCalledOnce();
    });

    it("throws on empty name", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.createEvalRule("", "eval-prompt")).rejects.toThrow("name is required");
    });

    it("throws on empty prompt_name", async () => {
      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      await expect(client.createEvalRule("helpfulness", "")).rejects.toThrow(
        "prompt_name is required"
      );
    });

    it("includes prompt_label in payload when provided", async () => {
      let capturedBody: Record<string, unknown> = {};
      const fetchSpy = mockFetch(async (_url, init) => {
        capturedBody = JSON.parse((init?.body as string) ?? "{}");
        return createResponse(201, {
          rule_id: "rule-456",
          name: "helpfulness",
          judge_model: "gpt-4o-mini",
          prompt_name: "eval-prompt",
          prompt_version: 3,
          prompt_label: "production",
          filter: {},
          sample_rate: 1.0,
          enabled: true,
        });
      });

      const client = new OmnevalClient({ baseUrl: "http://localhost:3000" });
      const result = await client.createEvalRule("helpfulness", "eval-prompt", {
        prompt_version: 3,
        prompt_label: "production",
      });

      expect(capturedBody.prompt_version).toBe(3);
      expect(capturedBody.prompt_label).toBe("production");
      expect(result.prompt_label).toBe("production");
      expect(fetchSpy).toHaveBeenCalledOnce();
    });
  });
});
